qiskit/circuit/qpy_serialization.py

```diff
@@ -0,0 +1,981 @@
+
+    This function is used to load a serialized QPY circuit file and create
+    :class:`~qiskit.circuit.QuantumCircuit` objects from its contents.
+    For example:
+
+    .. code-block:: python
+
+        from qiskit.circuit import qpy_serialization
+
+        with open('bell.qpy', 'rb') as fd:
+            circuits = qpy_serialization.load(fd)
+
+    or with a gzip compressed file:
+
+    .. code-block:: python
+
+        import gzip
+        from qiskit.circuit import qpy_serialization
+
+        with gzip.open('bell.qpy.gz', 'rb') as fd:
+            circuits = qpy_serialization.load(fd)
+
+    which will read the contents of the qpy and return a list of
+    :class:`~qiskit.circuit.QuantumCircuit` objects from the file.
+
+    Args:
+        file_obj (File): A file like object that contains the QPY binary
+            data for a circuit
+    Returns:
+        list: A list of the QuantumCircuit objects from the QPY data. A list
+            is always returned, even if there is only 1 circuit in the QPY
+            data.
+    Raises:
+        QiskitError: if ``file_obj`` is not a valid QPY file
+    """
+    file_header_raw = file_obj.read(FILE_HEADER_SIZE)
+    file_header = struct.unpack(FILE_HEADER_PACK, file_header_raw)
+    if file_header[0].decode("utf8") != "QISKIT":
+        raise QiskitError("Input file is not a valid QPY file")
+    version_parts = [int(x) for x in __version__.split(".")[0:3]]
+    header_version_parts = [file_header[2], file_header[3], file_header[4]]
+    if (
+        version_parts[0] < header_version_parts[0]
+        or (
+            version_parts[0] == header_version_parts[0]
+            and header_version_parts[1] > version_parts[1]
+        )
+        or (
+            version_parts[0] == header_version_parts[0]
+            and header_version_parts[1] == version_parts[1]
+            and header_version_parts[2] > version_parts[2]
+        )
+    ):
+        warnings.warn(
+            "The qiskit version used to generate the provided QPY "
+            "file, %s, is newer than the current qiskit version %s. "
+            "This may result in an error if the QPY file uses "
+            "instructions not present in this current qiskit "
+            "version" % (".".join(header_version_parts), __version__)
+        )
+    circuits = []
+    for _ in range(file_header[5]):
+        circuits.append(_read_circuit(file_obj))
+    return circuits
+
+
+def _read_circuit(file_obj):
+    header, name, metadata = _read_header(file_obj)
+    registers = {}
+    if header[5] > 0:
+        circ = QuantumCircuit(name=name, global_phase=header[1], metadata=metadata)
+        registers = _read_registers(file_obj, header[5])
+        for qreg in registers["q"].values():
+            min_index = min(qreg["index_map"].keys())
+            qubits = [Qubit() for i in range(min_index - len(circ.qubits))]
+            if qubits:
+                circ.add_bits(qubits)
+            circ.add_register(qreg["register"])
```

Comment:
Solving this generally I think could be done by first building the circuit with the right `num_qubits`/`num_clbits`, building registers with the bits found at the corresponding indices in `.qubits`/`.clbits`, and then adding those to the circuit. This can be done as a follow up, but a comment here explaining the intent (and current limitation) would be helpful.

Reply:
So I gave this a try and the issue is that there is a mismatch between bits from a standalone created register (ie `QuantumRegister(2, 'qr')`) vs one created from existing bits (ie `QuantumRegister(name='qr', bits=[Qubit() for _ in range(2)])`). So comparing the circuits on either end of the serialization weren't equal. I think we can fix this with a flag for the register struct that says whether it's standalone or not. But I think we should save that for the v2 format to demonstrate the backwards compatibility story, So I added a comment for now in: https://github.com/Qiskit/qiskit-terra/pull/5578/commits/4fca3deadfe8ae91ea196586ec715e99f212a0a1